File: code/day4.jl
function read_batch(file)
	lines = readlines(file)
	passports = [Dict{String, String}()]
	for line in lines
		if line == ""
			push!(passports, Dict())
		else
			for value in split(line, " ")
				pair = split(value, ":")
				last(passports)[pair[1]] = pair[2]
			end
		end
	end
	passports
end

function validate_passports(passports)
	pz1, pz2 = 0, 0
	for passport in passports
		fields = ["byr", "iyr", "eyr", "hgt", "ecl", "hcl", "pid"]
		if any([!haskey(passport, i) for i in fields])
			pz1 += 0; pz2 += 0
		else
			pz1 += 1
			if all([check_fields(passport, i) for i in fields])
				pz2 += 1
			else
				pz2 += 0
			end
		end
	end
	pz1, pz2
end

function check_fields(passport::Dict, field)
	if !haskey(passport, field)
		return false
	end

	value = get(passport, field, false)

	if field == "byr"
		return 1920 ≤ parse(Int, value) ≤ 2020
	elseif field == "iyr"
		return 2010 ≤ parse(Int, value) ≤ 2020
	elseif field == "eyr"
		return 2020 ≤ parse(Int, value) ≤ 2030
	elseif field == "hgt"
		m = match(r"(\d+)(in|cm)", value)
		if m == nothing
			return false
		end
		ht, ms = match(r"(\d+)(in|cm)", value).captures
		if ms == "in"
			return 59 ≤ parse(Int, ht) ≤ 76
		else
			return 150 ≤ parse(Int, ht) ≤ 193
		end
	elseif field == "ecl"
		return value ∈ ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
	elseif field == "hcl"
		return occursin(r"^#[0-9a-f]{6}$", value)
	elseif field == "pid"
		return occursin(r"^[0-9]{9}$", value)
	end
end

passports = read_batch("data/day4")
validate_passports(passports)

File: code/day4.R
read_batch <- function(file) {
  batch <- readLines(file)
  id <- 1
  ids <- rep(NA, length(batch))
  for (i in seq_along(batch)) {
    id <- id + ifelse(batch[i] == "", 1, 0)
    ids[i] <- id
  }
  lapply(split(batch, ids), function(x) trimws(paste(x, collapse = " ")))
}

get_param <- function(field) {
  sub(":.*", "", field)
}

check_byr <- function(byr) {
  yr <- as.integer(sub(".*:", "", byr))
  yr >= 1920L && yr <= 2002L
}

check_iyr <- function(iyr) {
  yr <- as.integer(sub(".*:", "", iyr))
  yr >= 2010L && yr <= 2020L
}

check_eyr <- function(eyr) {
  yr <- as.integer(sub(".*:", "", eyr))
  yr >= 2020L && yr <= 2030L
}

check_hgt <- function(hgt) {
  if (!grepl("cm|in", hgt)) {
    return(FALSE)
  }
  ms <- sub(".*:", "", hgt)
  ht <- as.integer(sub("cm|in", "", ms))
  if (grepl("cm", ms)) {
    return(ht >= 150L && ht <= 193L)
  }
  ht >= 59L && ht <= 76L
}

check_hcl <- function(hcl) {
  cl <- sub(".*:", "", hcl)
  if (!grepl("#", cl)) {
    return(FALSE)
  }
  col <- sub("#", "", cl)
  if (gsub("[0-9]|[a-f]", "", col) != "" || nchar(col) != 6) {
    return(FALSE)
  }
  TRUE
}

check_ecl <- function(ecl) {
  col <- c("amb", "blu", "brn", "gry", "grn", "hzl", "oth")
  sum(unlist(lapply(col, function(x) grepl(x, ecl)))) == 1
}

check_pid <- function(pid) {
  id <- sub(".*:", "", pid)
  if (gsub("[0-9]", "", id) != "" || nchar(id) != 9) {
    return(FALSE)
  }
  TRUE
}

valid_passports <- function(passports) {
  fields <- c("byr:", "iyr:", "eyr:", "hgt:", "hcl:", "ecl:", "pid:")
  p1 <- p2 <- rep(NA, length(passports))
  for (i in seq_along(passports)) {
    p1[i] <- all(unlist(lapply(fields, function(x) grepl(x, passports[i]))))
    if (p1[i]) {
      params <- unlist(strsplit(passports[[i]], " "))
      p2[i] <- all(unlist(lapply(params, function(x) {
        switch(sub(":.*", "", x), 
               byr = check_byr(x), 
               iyr = check_iyr(x), 
               eyr = check_eyr(x), 
               hgt = check_hgt(x), 
               hcl = check_hcl(x), 
               ecl = check_ecl(x), 
               pid = check_pid(x))
      })))
    } else {
      p2[i] <- FALSE
    }
  }
  cat("Puzzle 1:", sum(p1), "\n")
  cat("Puzzle 2:", sum(p2), "\n")
}

valid_passports(read_batch("data/day4"))
